### PR TITLE
feat: abitype

### PR DESCRIPTION
**Pull request by tmm**
## Description

Adds type-safety to contract interactions and [EIP-712](https://eips.ethereum.org/EIPS/eip-712) typed data. Most notably: Autocomplete and type checking for contract function names, arguments, return values, etc.

### Some examples in action

- [Demo at ETHGlobal Dev Tools Summit](https://youtu.be/Yd83o-9lW3o?t=1043)
- [Contract interaction](https://twitter.com/awkweb/status/1555678944770367493)
- [EIP-712 Typed Data](https://twitter.com/awkweb/status/1567671175345496064)

## Additional Information

- [x] I read the [contributing docs](/wagmi-dev/wagmi/blob/main/.github/CONTRIBUTING.md) (if this is your first contribution)

Your ENS/address: awkweb.eth

### CODE REVIEW

File: docs/package.json [hunk 1]
```diff
@@ -34,6 +34,6 @@
     "eslint-config-next": "^12.1.6",
     "postcss": "^8.4.13",
     "tailwindcss": "^3.0.24",
-    "typescript": "^4.6.4"
+    "typescript": "^4.7.4"
```
Comment (tmm) [inline docs/package.json:37]: min version for `abitype`

**Comment by tmm** [comment]
@SocketSecurity ignore abitype@0.0.17

File: packages/core/src/actions/contracts/readContracts.ts [hunk 2]
```diff
@@ -8,37 +14,116 @@ import {
...
+  : TContracts extends [infer Head, ...infer Tail]
+  ? ContractsConfig<[...Tail], [...Result, GetConfig<Head>], [...Depth, 1]>
+  : unknown[] extends TContracts
+  ? TContracts
+  : TContracts extends ReadContractsContractConfig<
+      infer TAbi,
+      infer TFunctionName
+    >[]
+  ? ReadContractsContractConfig<TAbi, TFunctionName>[]
+  : ReadContractsContractConfig[]
```
Comment (tmm) [inline packages/core/src/actions/contracts/readContracts.ts:81]: Was going to re-use this type logic from `multicall`, but the configuration is subtly different (multicall has `chainId` as a top level option and `readContracts` places `chainId` inside the contract object).

File: packages/core/src/constants/abis.ts [hunk 3]
```diff
@@ -1,36 +1,530 @@
...
-  'function allowance(address _owner, address _spender) public view returns (uint256 remaining)',
-  'function approve(address _spender, uint256 _value) public returns (bool success)',
-  'function balanceOf(address _owner) public view returns (uint256 balance)',
-  'function decimals() public view returns (uint8)',
-  'function name() public view returns (string)',
-  'function symbol() public view returns (string)',
-  'function totalSupply() public view returns (uint256)',
-  'function transfer(address _to, uint256 _value) public returns (bool success)',
-  'function transferFrom(address _from, address _to, uint256 _value) public returns (bool success)',
-]
```
Comment (tmm) [inline packages/core/src/constants/abis.ts:14]: Changed these from ethers' human-readable format to standard ABI format so they work with type inference. (Could parse the human-readable format with TypeScript, but I'll save that for another day 😁.)

File: packages/core/src/types/index.ts [hunk 4]
```diff
@@ -7,7 +8,19 @@ import {
...
 
-export type Address = `0x${string}`
+declare module 'abitype' {
+  export interface Config {
+    // TODO: Drop `BigNumber` once ethers supports `bigint` natively
+    BigIntType: BigNumber
+    BytesType: `0x${string}` | ArrayLike<number>
+    IntType: number
+  }
+}
```
Comment (tmm) [inline packages/core/src/types/index.ts:18]: Configure ABIType to work with ethers types

File: packages/react/src/hooks/accounts/useSignTypedData.ts [hunk 5]
```diff
@@ -55,14 +73,22 @@ export function useSignTypedData({
...
-    (args?: SignTypedDataArgs) =>
-      mutate(args || <SignTypedDataArgs>{ domain, types, value }),
+    (args?: Partial<SignTypedDataArgs<TTypedData, TSchema>>) =>
+      mutate(<SignTypedDataArgs<TTypedData, TSchema>>{
+        domain: args?.domain ?? domain,
+        types: args?.types ?? types,
+        value: args?.value ?? value,
+      }),
     [domain, types, value, mutate],
   )
```
Comment (tmm) [inline packages/react/src/hooks/accounts/useSignTypedData.ts:83]: Allows you to pass partial values: `signTypedData({ value })`

File: packages/react/src/hooks/contracts/useContractEvent.ts [hunk 6]
```diff
@@ -1,44 +1,67 @@
...
   })
-  const listenerRef = React.useRef(listener)
-  listenerRef.current = listener
+  const callbackRef = React.useRef(callback)
+  callbackRef.current = callback
 
   React.useEffect(() => {
-    const handler = (...event: Array<Parameters<Contract['on']>[1]>) =>
-      listenerRef.current(event)
+    const handler = (...event: any[]) => callbackRef.current(...event)
```
Comment (tmm) [inline packages/react/src/hooks/contracts/useContractEvent.ts:64]: wasn't spreading params into callback before! caught while testing in the playground with type-safety!

File: packages/core/src/types/utils.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        ? void
+        : TLength extends 1
+        ? AbiParameterToPrimitiveType<TOutputs[0]>
+        : // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        TOutputs extends readonly [...infer _]
+        ? {
+            [Component in TOutputs[number] as Component['name'] extends ''
+              ? never
+              : Component['name']]: AbiParameterToPrimitiveType<Component>
+          } & AbiParametersToPrimitiveTypes<TOutputs>
```
Comment (tmm) [inline packages/core/src/types/utils.ts:54]: @jxom lmk how you want to proceed with typing here. this currently sets the response to the array-object hybrid that ethers returns.

Comment (jxom) [reply packages/core/src/types/utils.ts:54]: This seems fine!

File: packages/react/src/hooks/contracts/useContractReads.test.ts [hunk 8]
```diff
@@ -6,24 +7,24 @@ import {
...
+    args: ['0x27a69ffba1e939ddcfecc8c7e0f967b872bac65c'],
   },
   {
     ...wagmigotchiContractConfig,
     functionName: 'love',
-    args: '0xa5cc3c03994DB5b0d9A5eEdD10CabaB0813678AC',
+    args: ['0xa5cc3c03994DB5b0d9A5eEdD10CabaB0813678AC'],
   },
-  { ...wagmigotchiContractConfig, functionName: 'getAlive' },
+  { ...wagmigotchiContractConfig, functionName: 'getAlive', args: [] },
```
Comment (jxom) [inline packages/react/src/hooks/contracts/useContractReads.test.ts:23]: From internal convo: do we need to pass an empty `args` here? Type error seems to disappear if we add `'getAlive' as const`.

File: packages/core/src/actions/contracts/readContracts.test.ts [hunk 9]
```diff
@@ -1,3 +1,4 @@
+import { BigNumber } from 'ethers'
```
Comment (jxom) [inline packages/core/src/actions/contracts/readContracts.test.ts:1]: On L48, the type of `await readContracts({ contracts })` resolves to `any[]`. Same goes for the `useContractReads` & `useContractInfiniteReads` tests. Should these be typed?

File: packages/react/src/hooks/contracts/useContractInfiniteReads.ts [hunk 10]
```diff
@@ -1,6 +1,6 @@
 import {
```
Comment (jxom) [inline packages/react/src/hooks/contracts/useContractInfiniteReads.ts:1]: When I consume this hook, it seems that the page items are of `unknown`.

```tsx
const { data } = useContractInfiniteReads({
  //     ^ `InfiniteData<[unknown]> | undefined`
  cacheKey: 'contracts',
  contracts: (index = 0) => [
    {
      addressOrName: '0x1dfe7ca09e99d10835bf73044a23b73fc20623df',
      contractInterface: mlootContractInterface,
      functionName: 'getChest',
      args: [index],
    },
  ],
  getNextPageParam: (_, pages) => pages.length + 1,
})
```

File: packages/core/src/types/utils.ts [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+  TFunction extends AbiFunction & { type: 'function' },
+> = TFunction['inputs'] extends infer TInputs extends readonly any[]
+  ? Or<IsNever<TInputs>, NotEqual<TAbi, Abi>> extends true
+    ? {
+        /**
+         * Arguments to pass contract method
+         *
+         * Use a [const assertion](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions) on {@link contractInterface} for type inference.
+         */
+        args?: readonly any[]
```
Comment (jxom) [inline packages/core/src/types/utils.ts:25]: Should this be `any | readonly any[]` for backwards compat? Or is there a reason why it has to be an array?

Comment (jxom) [reply packages/core/src/types/utils.ts:25]: Yeah, fair call. Definitely prefer strictly passing an array!

File: packages/react/src/hooks/contracts/useContractInfiniteReads.ts (hunk 10 above)
Comment (tmm) [reply packages/react/src/hooks/contracts/useContractInfiniteReads.ts:1]: should be fixed now!

File: docs/pages/docs/typescript.en-US.mdx [hunk 12]
```diff
@@ -0,0 +1,55 @@
+---
```
Comment (tmm) [inline docs/pages/docs/typescript.en-US.mdx:1]: still wip

File: packages/react/src/hooks/contracts/useContract.test.ts [hunk 13]
```diff
@@ -28,11 +28,29 @@ describe('useContract', () => {
...
-      expect(result.current.provider).toBeNull()
+      expect(result.current?.provider).toBeNull()
 
       signerOrProvider = getProvider()
       rerender()
-      expect(result.current.provider).not.toBeNull()
+      expect(result.current?.provider).not.toBeNull()
+    })
+
+    it('no adressOrName', async () => {
```
Comment (jxom) [inline packages/react/src/hooks/contracts/useContract.test.ts:38]:
```suggestion
    it('no addressOrName', async () => {
```

File: packages/react/src/hooks/contracts/useContractEvent.ts [hunk 14]
```diff
@@ -1,51 +1,110 @@
...
+export function useContractEvent<
+  TAbi extends Abi | readonly unknown[],
+  TEventName extends string,
+>({
   addressOrName,
   chainId,
   contractInterface,
   listener,
   eventName,
-  signerOrProvider,
```
Comment (jxom) [inline packages/react/src/hooks/contracts/useContractEvent.ts:26]: Can we note this as a breaking change in the changesets?

File: packages/react/src/hooks/contracts/useContractRead.ts [hunk 15]
```diff
@@ -5,26 +5,39 @@ import {
...
-}
-
-export type UseContractReadConfig = QueryConfig<ReadContractResult, Error>
+export type UseContractReadConfig<
+  TAbi = Abi,
+  TFunctionName = string,
+> = ReadContractConfig<
+  TAbi,
+  TFunctionName,
+  { isAddressOptional: true; isArgsOptional: true }
```
Comment (jxom) [inline packages/react/src/hooks/contracts/useContractRead.ts:21]: Are we using these?

File: packages/react/src/hooks/contracts/useContractRead.ts [hunk 16]
```diff
@@ -103,10 +122,10 @@ export function useContractRead({
   )
 
   const enabled = React.useMemo(() => {
-    let enabled = Boolean(enabled_ && addressOrName && functionName)
```
Comment (jxom) [inline packages/react/src/hooks/contracts/useContractRead.ts:106]: How come we omitted functionName here?

Comment (tmm) [reply packages/react/src/hooks/contracts/useContractRead.ts:106]: `functionName` is a required config option so seemed unnecessary. think we should include it?

File: packages/react/src/hooks/contracts/useContractRead.test.ts [hunk 17]
```diff
@@ -1,7 +1,10 @@
+import { ResolvedConfig } from 'abitype'
```
Comment (jxom) [inline packages/react/src/hooks/contracts/useContractRead.test.ts:1]: Should we add tests for optional config (`addressOrName`, `args`)?

File: packages/react/src/hooks/contracts/useContractReads.test.ts [hunk 18]
```diff
@@ -1,31 +1,34 @@
+import { ResolvedConfig } from 'abitype'
```
Comment (jxom) [inline packages/react/src/hooks/contracts/useContractReads.test.ts:1]: Should we add tests for optional config (addressOrName, args)?

File: packages/react/src/hooks/contracts/useContractWrite.test.ts [hunk 19]
```diff
@@ -1,3 +1,5 @@
+import { Abi, ExtractAbiFunctionNames } from 'abitype'
```
Comment (jxom) [inline packages/react/src/hooks/contracts/useContractWrite.test.ts:1]: Should we add tests for optional config (addressOrName, args)?

File: packages/react/src/hooks/contracts/usePrepareContractWrite.test.ts [hunk 20]
```diff
@@ -1,3 +1,5 @@
+import { Abi } from 'abitype'
```
Comment (jxom) [inline packages/react/src/hooks/contracts/usePrepareContractWrite.test.ts:1]: Should we add tests for optional config (addressOrName, args)?

**Review by jxom — APPROVED** [review]
Looking really solid 🔥

**Comment by holic** [comment]
🎉🎉🎉

**Pull request merged**